### PR TITLE
WIP: Added Zilionis dataset

### PR DESCRIPTION
**Pull request by jenzopr**
Hi Aaron,

this should be considered WIP. I started with all cells from the Zilionis lung cancer data set for mouse and human. I used baron pancreas as template. Can you check if there are major issues so far?

A few questions from my side:
- what/how to generate the bib entry? If I generate a bibtex entry, it looks very different compared to the existing
- How to upload to Experimenthub? Do you do that?
- Additional metadata will be implemented in `inst/scripts/make-zilionis-lung-data.Rmd` and only the subset implemented in the getter function? Did I get this right from our email conversation?

Best,
Jens

### CODE REVIEW

**Comment by LTLA** [comment]
I recently made explicit directories for each version, so yours would go into 2.4.0 for both the `extdata` and `scripts` (it's just easier to sync this to the ExperimentHub directories).

> what/how to generate the bib entry? If I generate a bibtex entry, it looks very different compared to the existing

I usually generate the bib entries with https://www.bioinformatics.org/texmed/.

> How to upload to Experimenthub? Do you do that?

Yes, I'll do that. Slightly annoying because you won't be able to test your R function until I do so, but just do your best.

> Additional metadata will be implemented in inst/scripts/make-zilionis-lung-data.Rmd and only the subset implemented in the getter function? Did I get this right from our email conversation?

That's generally correct, I'll have a look at your specific code later.

File: inst/scripts/make-zilionis-lung-data.Rmd [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+```{r}
+stopifnot(length(unique(lapply(all.human, function(x) rownames(x$mat))))==1L)
+counts <- do.call(cbind, lapply(all.human, function(x) x$mat))
+dim(counts)
+```
+
+We do the same thing with the column metadata.
+
+```{r}
+samples <- sapply(stringi::stri_split_fixed(hs.files, pattern = "_"), function(x) x[3])
```
Comment (LTLA) [inline inst/scripts/make-zilionis-lung-data.Rmd:99]: I'm sure you'd survive with `vapply(strsplit(hs.files, "_"), "[", i=3, "")`.

File: inst/scripts/make-zilionis-lung-data.Rmd [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+barcode <- lapply(all.human, function(x) x$label)
+origin <- rep(samples, times = sapply(barcode, length))
```
Comment (LTLA) [inline inst/scripts/make-zilionis-lung-data.Rmd:101]: `lengths(barcode)`.

File: inst/scripts/make-zilionis-lung-data.Rmd [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+donor <- stringi::stri_match_first_regex(origin, "p[0-9]+")[,1]
+replicate <- stringi::stri_match_first_regex(origin, "[t|b][0-9]+")[,1]
```
Comment (LTLA) [inline inst/scripts/make-zilionis-lung-data.Rmd:103]: A simpler approach would be `sub("[bt].*", "", origin)` and also `sub("p.", "", origin)`.

File: inst/scripts/make-zilionis-lung-data.Rmd [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+```
+
+# Reading in human data
+
+We set up a function to load in each set of counts as a sparse matrix.
+
+```{r}
+library(Matrix)
+FUN <- function(X) {
+    input <- read.csv(X, sep="\t", stringsAsFactors=FALSE)
```
Comment (LTLA) [inline inst/scripts/make-zilionis-lung-data.Rmd:45]: You can set `row.names=1` to save yourself the need to subset away the first column.

File: inst/scripts/make-zilionis-lung-data.Rmd [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+stopifnot(length(unique(lapply(all.mouse, function(x) rownames(x$mat))))==1L)
+counts <- do.call(cbind, lapply(all.mouse, function(x) x$mat))
+dim(counts)
+```
+
+We do the same thing with the column metadata.
+
+```{r}
+animal <- sapply(stringi::stri_split_fixed(mm.files, pattern = "_"), function(x) paste0(x[3], x[4]))
+replicate <- sapply(stringi::stri_split_fixed(mm.files, pattern = "_"), function(x) x[5])
```
Comment (LTLA) [inline inst/scripts/make-zilionis-lung-data.Rmd:158]: See above comments about `strsplit()`.

**Comment by jenzopr** [comment]
Hi Aaron again!

I added additional metadata for 54773 human and 15939 mouse cells from the supplementary files available for GSE127465. It contains 28 and 12 additional columns for `colData`, which noch looks something like

```
DataFrame with 17549 rows and 15 columns
       library   animal replicate   tissue     barcode Tumor.or.healthy Biological.replicate Library.prep.batch Total.counts
      <factor> <factor>  <factor> <factor> <character>         <factor>             <factor>           <factor>    <numeric>
1        h_1_1      h_1         1  healthy      bc0001                h            Healthy 1    round1_20151128        11915
2        h_1_1      h_1         1  healthy      bc0002                h            Healthy 1    round1_20151128        11675
3        h_1_1      h_1         1  healthy      bc0003                h            Healthy 1    round1_20151128        12536
4        h_1_1      h_1         1  healthy      bc0004                h            Healthy 1    round1_20151128        12499
5        h_1_1      h_1         1  healthy      bc0005                h            Healthy 1    round1_20151128         9586
...        ...      ...       ...      ...         ...              ...                  ...                ...          ...
17545    t_2_4      t_2         4    tumor      bc1084                t      Tumor-bearing 2    round3_20160313          912
17546    t_2_4      t_2         4    tumor      bc1085                t      Tumor-bearing 2    round3_20160313          926
17547    t_2_4      t_2         4    tumor      bc1086                t      Tumor-bearing 2    round3_20160313          795
17548    t_2_4      t_2         4    tumor      bc1087               NA                   NA                 NA           NA
17549    t_2_4      t_2         4    tumor      bc1088                t      Tumor-bearing 2    round3_20160313          661
      Percent.counts.from.mitochondrial.genes Most.likely.Immgen.cell.type Major.cell.type Minor.subset             x              y
                                    <numeric>                     <factor>        <factor>     <factor>     <numeric>      <numeric>
1                                        0.21                 GN_Arth_SynF     Neutrophils           N1 262.182565585 -776.218780693
2                                        0.24                 GN_Arth_SynF     Neutrophils           N1 141.317902324 -766.314485969
3                                       2.728                  MF_PPAR-_Lu         MoMacDC         Mac4 1303.09815838  69.2643250724
4                                       2.728                        MF_Lu         MoMacDC         Mac4 1299.30882847  109.660673297
5                                       2.378                        MF_Lu         MoMacDC         Mac4 1305.98511809  66.6067642274
...                                       ...                          ...             ...          ...           ...            ...
17545                           3.83771929825                      B_Fo_Sp         B cells      B cells 240.758903983  292.600115176
17546                           5.50755939525                   GN_UrAc_PC     Neutrophils           N4 1037.34864874 -796.034347295
17547                           0.62893081761                 GN_Arth_SynF     Neutrophils           N3 413.387189669 -650.431907737
17548                                      NA                           NA              NA           NA            NA             NA
17549                           8.01815431165                     B_Fo_MLN         B cells      B cells 208.024524033  298.377908521
```

Whats your opinion on removing duplicated columns, and selecting only meaningful / helpful columns, like `Most.likely.Immgen.cell.type`, `Major.cell.type` and `Minor.subset`? `x`and `y` are coordinates from a 2D Spring map. We could include them in `reducedDims`?

**Comment by LTLA** [comment]
> Whats your opinion on removing duplicated columns 

Yes. Have a look at the automated cleaning of the ArrayExpress metadata here:

https://github.com/LTLA/scRNAseq/blob/caeefbff728bddfe0bb41ba18298acd3c4cf3110/inst/scripts/2.0.0/make-richard-tcell-data.Rmd#L53-L69

> We could include them in reducedDims?

Yes, but better to do that unpacking in the getter function, avoid storing another thing in EHub.

**Comment by LTLA** [comment]
What's the status?

**Comment by jenzopr** [comment]
Other projects :-/ I will try to finish this tomorrow. Is this okay for you?

**Comment by LTLA** [comment]
Further comments:

- Make sure your `metadata.R` file writes to `extdata/2.4.0/` and uses 2.4.0 in the `RDataPath`. Also replace `DataFrame` with `DFrame`, which is the new name of the class.
- Do me a favor and move all your Zilionis files to `scripts/2.4.0/`, see the current `master`.
- Not on my watch:

  ```r
  read.table(tarball, sep = "\t", header = T) %>% 
    dplyr::left_join(coldata, ., by = c("library" = "Library", "barcode" = "Barcode")) %>% 
    DataFrame(.) -> colData
  ```
  
  Hell no. `base::merge` instead of **dplyr**, no dots, no pipes. Use `check.names=FALSE` in `read.table()` to get full names, and use `TRUE` instead of `T`. 
- You threw out `Minor subset` in the human annotation. Dunno if that was important.

**Comment by jenzopr** [comment]
I tried to incorporate the changes. Is `DFrame` new in R4.0.0?

**Comment by LTLA** [comment]
`DFrame` is just the class name for the object constructed by `DataFrame`.

**Comment by jenzopr** [comment]
Any further wishes? :-)

**Comment by LTLA** [comment]
Good enough. Let's do this.

**Comment by LTLA** [comment]
@jenzopr It is done. You should see that I have made some further edits to the column metadata: have a look at it to check that everything is in order.

**Comment by jenzopr** [comment]
Nice! I will do so!

**Pull request merged**